File: AoC2020/solutions/05/solution.jl
#Advent of Code 2020 day 5
input_path = (@__DIR__) * "/input"

function load(file_path=input_path)
    # Read data from file
    data = read((@__DIR__) * "/input", String)
    # Split into lines
    lines = split(strip(data), "\n")
    return lines
end

function calc_data(data)
    # Transforms the string of FBLR to binary number
    transform = Dict('F'=>'0', 'B'=>'1', 'L'=>'0', 'R'=>'1')
    data = [map(x -> transform[x], pid) for pid in data]
    # Parses binary to int and sorts the array
    return sort(parse.(Int, data; base=2))
end

function task1(data)
    # Max value of the data
    return data[end]
end

function task2(data)
    # Only value from [min, max] that is not present in the data
    return filter(x -> !(x in data), data[1]:data[end])[1]
end

function solution()
    data = load()
    #println(data)
    data = calc_data(data)
    println("First:  $(task1(data))")
    println("Second: $(task2(data))")
end

solution()

File: AoC2020/solutions/05/solution.py
#Advent of Code 2020 day 5
from util import *
import re
DAY = 5


def get_data():
    return input_lines(DAY)


def get_seat_long(pid):
    # Turn pid to binary
    pid = re.sub(r"[FL]", '0', pid)
    pid = re.sub(r"[BR]", '1', pid)

    # Calculate row
    row = int(pid[0:7], 2)
    # Calculate column
    column = int(pid[7:11], 2)
    # Result
    return row*8+column


def get_seat(pid):
    # Pid is just a binary representation of the seat location
    return int(re.sub(r"[BR]", '1', re.sub(r"[FL]", '0', pid)), 2)


def solution():
    data = get_data()

    ids = sorted([get_seat(x) for x in data])
    first = ids[-1]
    second = next(filter(lambda x: x not in ids, range(ids[0], ids[-1])))

    print(f"First:  {first}")
    print(f"Second: {second}")


def golf():
    d = input_lines(DAY)
    i = {int(x.translate("".maketrans("FBLR", "0101")), 2) for x in d}
    print(f"First:  {max(i)}")
    print(f"Second: {min(set(range(min(i), max(i)))-i)}")


solution()
golf()

